docs/design.md

```diff
@@ -0,0 +1,371 @@
+   data from it instantaneously, because (a) an index exists and is loaded, and
+   (b) data is locally accessible (e.g. it doesn't need to be fetched from a
+   remote mount).
+2. **Unavailable:** the system knows about this shard, but is not capable of
+   serving data from it because the shard is being initialized, or the mount is
+   not available locally, but still accessible with work (e.g. fetching from a
+   remote location).
+3. **Destroyed**: the shard is no longer available; this is permanent condition.
+
+### Operations
+
+#### Shard registration
+
+To register a new shard, call:
+
+```go
+dagst.RegisterShard(key []byte, mount Mount, opts ...RegisterOpts) error
+```
+
+1. This method takes a shard key and a `Mount`.
+2. It initializes the shard in `Unavailable` state.
+3. Calls `mount.Info()` to determine if the mount is of local or remote type.
+4. Calls `mount.Stat()` to determine if the mount target exists. If not, it
+   errors the registration.
+5. If remote, it fetches the remote resource into the scrap area.
+6. It determines the kind of CAR it is, and populates the index accordingly.
+7. Sets the state to `Available`.
+8. Returns.
+
+This method is _synchronous_. It returns only when the shard is fully indexed
+and available for serving data. This embodies the consistency property of the
+ACID model, whereby an INSERT is immediately queriable upon return.
+
+_RegisterOpts is an extension point._ In the future it can be used to pass in an
+unseal/decryption function to be used on access (e.g. such as when fast, random
+unseal is available).
+
+#### Shard destruction
+
+To destroy a shard, call:
+
+```go
+dagst.DestroyShard(key []byte) (destroyed bool, err error)
+```
+
+#### Other operations
+
+  * `[Pin/Unpin]Shard()`: retains the shard data in the local scrap area.
+  * `ReleaseShard()`: dispose of / release local scrap copies or other resources
+    on demand (e.g. unsealed copy).
+  * `[Lock/Unlock]Shard()`: for exclusive access.
+  * `[Incr/Decr]Shard()`: refcounting on shards.
+
+### Mounts
+
+Shards can be located anywhere, and can come and go dynamically e.g. Filecoin
+deals expire, removable media is attached/detached, or the IPFS user purges
+content.
+
+It is possible to mount shards with CARs accessible through the local
+filesystem, detachable mounts, NFS mounts, distributed filesystems like
+Ceph/GlusterFS, HTTP, FTP, etc.
+
+This versatility is provided by an abstraction called `mount.Mount`, a pluggable
+component which encapsulates operations/logic to:
+
+1. `Fetch() (io.ReadCloser, error)` Load a CAR from its origin.
+2. `Info() mount.Info` Provides info about the mount, e.g. whether it's local or
+   remote. This is used to determine whether the fetched CAR needs to be copied
+   to a scrap area.
+3. `Stat() (mount.Stat, error)` Equivalent to a filesystem stat, provides
+   information about the target of the mount: whether it exists, size, etc.
+
```

Comment:
I think we should add the `Marshal` & `Unmarshal` API to the `mount.Mount` interface to make it easy for the DAGStore to implement Mount persistence.

Reply:
Yup, already taken into account in the skeleton, but in a slightly different manner (using a registry).